// angular-admin-lte/esm5/lib/color/color.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, Renderer2, ElementRef } from '@angular/core';
import { ColorService } from './color.service';
var BackgroundColorDirective = /** @class */ (function () {
    /**
     * @method constructor
     * @param elementRef   [description]
     * @param renderer2    [description]
     * @param colorService [description]
     */
    function BackgroundColorDirective(elementRef, renderer2, colorService) {
        this.elementRef = elementRef;
        this.renderer2 = renderer2;
        this.colorService = colorService;
        this.condition = true;
    }
    Object.defineProperty(BackgroundColorDirective.prototype, "color", {
        set: /**
         * @param {?} color
         * @return {?}
         */
        function (color) {
            this.colorService.setBackgroundColor(color, this.condition, this.property, this.prefix);
        },
        enumerable: true,
        configurable: true
    });
    BackgroundColorDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[mkColor]',
                    providers: [ColorService]
                },] },
    ];
    BackgroundColorDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: ColorService }
    ]; };
    BackgroundColorDirective.propDecorators = {
        condition: [{ type: Input, args: ['mkColorCondition',] }],
        prefix: [{ type: Input, args: ['mkColorPrefix',] }],
        property: [{ type: Input, args: ['mkColorProperty',] }],
        color: [{ type: Input, args: ['mkColor',] }]
    };
    return BackgroundColorDirective;
}());
export { BackgroundColorDirective };
if (false) {
    /** @type {?} */
    BackgroundColorDirective.prototype.condition;
    /** @type {?} */
    BackgroundColorDirective.prototype.prefix;
    /** @type {?} */
    BackgroundColorDirective.prototype.property;
    /**
     * @type {?}
     * @private
     */
    BackgroundColorDirective.prototype.elementRef;
    /**
     * @type {?}
     * @private
     */
    BackgroundColorDirective.prototype.renderer2;
    /**
     * @type {?}
     * @private
     */
    BackgroundColorDirective.prototype.colorService;
}
var ColorDirective = /** @class */ (function () {
    /**
     * @method constructor
     * @param elementRef   [description]
     * @param renderer2    [description]
     * @param colorService [description]
     */
    function ColorDirective(elementRef, renderer2, colorService) {
        this.elementRef = elementRef;
        this.renderer2 = renderer2;
        this.colorService = colorService;
    }
    Object.defineProperty(ColorDirective.prototype, "color", {
        set: /**
         * @param {?} color
         * @return {?}
         */
        function (color) {
            this.colorService.setFontColor(color);
        },
        enumerable: true,
        configurable: true
    });
    ColorDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[mkFontColor]',
                    providers: [ColorService]
                },] },
    ];
    ColorDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: ColorService }
    ]; };
    ColorDirective.propDecorators = {
        color: [{ type: Input, args: ['mkFontColor',] }]
    };
    return ColorDirective;
}());
export { ColorDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ColorDirective.prototype.elementRef;
    /**
     * @type {?}
     * @private
     */
    ColorDirective.prototype.renderer2;
    /**
     * @type {?}
     * @private
     */
    ColorDirective.prototype.colorService;
}
