// angular-admin-lte/esm5/lib/layout/footer/footer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChild, ViewChild, TemplateRef, ElementRef } from '@angular/core';
import { FooterService } from './footer.service';
/**
 * Footer Left
 */
var FooterLeftComponent = /** @class */ (function () {
    /**
     * Footer Left
     */
    function FooterLeftComponent() {
    }
    FooterLeftComponent.decorators = [
        { type: Component, args: [{
                    selector: 'mk-layout-footer-left',
                    template: '<ng-template #templateRef><ng-content></ng-content></ng-template>'
                },] },
    ];
    FooterLeftComponent.propDecorators = {
        templateRef: [{ type: ViewChild, args: ['templateRef',] }]
    };
    return FooterLeftComponent;
}());
export { FooterLeftComponent };
if (false) {
    /** @type {?} */
    FooterLeftComponent.prototype.templateRef;
}
/**
 * Footer Right
 */
var FooterRightComponent = /** @class */ (function () {
    /**
     * Footer Right
     */
    function FooterRightComponent() {
    }
    FooterRightComponent.decorators = [
        { type: Component, args: [{
                    selector: 'mk-layout-footer-right',
                    template: '<ng-template #templateRef><ng-content></ng-content></ng-template>'
                },] },
    ];
    FooterRightComponent.propDecorators = {
        templateRef: [{ type: ViewChild, args: ['templateRef',] }]
    };
    return FooterRightComponent;
}());
export { FooterRightComponent };
if (false) {
    /** @type {?} */
    FooterRightComponent.prototype.templateRef;
}
var FooterComponent = /** @class */ (function () {
    function FooterComponent(elementRef, footerService) {
        this.elementRef = elementRef;
        this.footerService = footerService;
    }
    /**
     * @return {?}
     */
    FooterComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.footerService.elementRef = this.elementRef;
    };
    FooterComponent.decorators = [
        { type: Component, args: [{
                    selector: 'mk-layout-footer',
                    template: "<footer class=\"main-footer\">\n  <div class=\"pull-right hidden-xs\">\n    <ng-template [ngTemplateOutlet]=\"footerRightComponent?.templateRef\"></ng-template>\n  </div>\n  <ng-template [ngTemplateOutlet]=\"footerLeftComponent?.templateRef\"></ng-template>\n</footer>\n",
                    styles: [":host{display:block}"]
                },] },
    ];
    FooterComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: FooterService }
    ]; };
    FooterComponent.propDecorators = {
        footerLeftComponent: [{ type: ContentChild, args: [FooterLeftComponent,] }],
        footerRightComponent: [{ type: ContentChild, args: [FooterRightComponent,] }]
    };
    return FooterComponent;
}());
export { FooterComponent };
if (false) {
    /** @type {?} */
    FooterComponent.prototype.footerLeftComponent;
    /** @type {?} */
    FooterComponent.prototype.footerRightComponent;
    /**
     * @type {?}
     * @private
     */
    FooterComponent.prototype.elementRef;
    /**
     * @type {?}
     * @private
     */
    FooterComponent.prototype.footerService;
}
