// angular-admin-lte/esm5/lib/box-small/box-small.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ContentChild } from '@angular/core';
import { BoxSmallContentDirective, BoxSmallFooterDirective, BoxSmallHeaderDirective } from './box-small.directive';
/*
 *
 */
var BoxSmallComponent = /** @class */ (function () {
    /*
     *
     */
    function BoxSmallComponent() {
        this.contentStyleClass = 'small-box-content';
        this.footerStyleClass = 'small-box-footer';
        this.headerStyleClass = 'small-box-header';
        this.iconStyleClass = 'ion ion-bag';
        this.styleClass = 'small-box';
    }
    BoxSmallComponent.decorators = [
        { type: Component, args: [{
                    selector: 'mk-box-small',
                    template: "<div [ngClass]=\"styleClass\" [mkColor]=\"backgroundColor\" mkColorProperty=\"background-color\">\n  <div class=\"inner\">\n    <h3 *ngIf=\"header || boxSmallHeaderDirective\" [ngClass]=\"headerStyleClass\" [mkFontColor]=\"headerColor\">\n      {{header}}\n      <ng-content select=\"mk-box-small-header\"></ng-content>\n    </h3>\n    <p [ngClass]=\"contentStyleClass\" [mkFontColor]=\"contentColor\">\n      <ng-container *ngIf=\"boxSmallHeaderDirective || boxSmallContentDirective || boxSmallFooterDirective; else noDirective\">\n        <ng-content select=\"mk-box-small-content\"></ng-content>\n      </ng-container>\n      <ng-template #noDirective>\n        <ng-content></ng-content>\n      </ng-template>\n    </p>\n  </div>\n  <div *ngIf=\"iconStyleClass\" class=\"icon\">\n    <i [ngClass]=\"iconStyleClass\" [mkFontColor]=\"iconColor\"></i>\n  </div>\n  <span *ngIf=\"footer || boxSmallFooterDirective\" [ngClass]=\"footerStyleClass\" [mkFontColor]=\"footerColor\">\n    {{footer}}\n    <ng-content select=\"mk-box-small-footer\"></ng-content>\n  </span>\n</div>\n",
                    styles: [".small-box.bg-color{color:#fff}/deep/ .small-box-footer:hover{cursor:pointer}/deep/ .small-box-footer a{color:rgba(255,255,255,.8)}/deep/ .small-box-footer:hover a{color:#fff}"]
                },] },
    ];
    BoxSmallComponent.propDecorators = {
        backgroundColor: [{ type: Input }],
        contentColor: [{ type: Input }],
        contentStyleClass: [{ type: Input }],
        footer: [{ type: Input }],
        footerColor: [{ type: Input }],
        footerStyleClass: [{ type: Input }],
        header: [{ type: Input }],
        headerColor: [{ type: Input }],
        headerStyleClass: [{ type: Input }],
        iconColor: [{ type: Input }],
        iconStyleClass: [{ type: Input }],
        styleClass: [{ type: Input }],
        boxSmallHeaderDirective: [{ type: ContentChild, args: [BoxSmallHeaderDirective,] }],
        boxSmallFooterDirective: [{ type: ContentChild, args: [BoxSmallFooterDirective,] }],
        boxSmallContentDirective: [{ type: ContentChild, args: [BoxSmallContentDirective,] }]
    };
    return BoxSmallComponent;
}());
export { BoxSmallComponent };
if (false) {
    /** @type {?} */
    BoxSmallComponent.prototype.backgroundColor;
    /** @type {?} */
    BoxSmallComponent.prototype.contentColor;
    /** @type {?} */
    BoxSmallComponent.prototype.contentStyleClass;
    /** @type {?} */
    BoxSmallComponent.prototype.footer;
    /** @type {?} */
    BoxSmallComponent.prototype.footerColor;
    /** @type {?} */
    BoxSmallComponent.prototype.footerStyleClass;
    /** @type {?} */
    BoxSmallComponent.prototype.header;
    /** @type {?} */
    BoxSmallComponent.prototype.headerColor;
    /** @type {?} */
    BoxSmallComponent.prototype.headerStyleClass;
    /** @type {?} */
    BoxSmallComponent.prototype.iconColor;
    /** @type {?} */
    BoxSmallComponent.prototype.iconStyleClass;
    /** @type {?} */
    BoxSmallComponent.prototype.styleClass;
    /** @type {?} */
    BoxSmallComponent.prototype.boxSmallHeaderDirective;
    /** @type {?} */
    BoxSmallComponent.prototype.boxSmallFooterDirective;
    /** @type {?} */
    BoxSmallComponent.prototype.boxSmallContentDirective;
}
