// angular-admin-lte/esm2015/lib/layout/content/content.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ChangeDetectionStrategy, ChangeDetectorRef, ViewChild, ElementRef } from '@angular/core';
import { Router, NavigationStart, NavigationEnd } from '@angular/router';
import { Title } from '@angular/platform-browser';
import { LayoutStore } from '../layout.store';
import { RoutingService } from '../../services/routing.service';
import { SidebarRightService } from '../sidebar-right/sidebar-right.service';
import { HeaderService } from '../header/header.service';
import { FooterService } from '../footer/footer.service';
import { removeSubscriptions } from '../../helpers';
export class ContentComponent {
    /**
     * \@method constructor
     * @param {?} layoutStore
     * @param {?} routingService
     * @param {?} titleService
     * @param {?} elementRef
     * @param {?} changeDetectorRef
     * @param {?} sidebarRightService
     * @param {?} headerService
     * @param {?} footerService
     * @param {?} router
     */
    constructor(layoutStore, routingService, titleService, elementRef, changeDetectorRef, sidebarRightService, headerService, footerService, router) {
        this.layoutStore = layoutStore;
        this.routingService = routingService;
        this.titleService = titleService;
        this.elementRef = elementRef;
        this.changeDetectorRef = changeDetectorRef;
        this.sidebarRightService = sidebarRightService;
        this.headerService = headerService;
        this.footerService = footerService;
        this.router = router;
        this.subscriptions = [];
    }
    /**
     * \@method ngOnInit
     * @return {?}
     */
    ngOnInit() {
        this.titleTag = this.titleService.getTitle();
        this.subscriptions.push(this.routingService.onChange.subscribe((value) => {
            if (value && value[value.length - 1]) {
                this.titleService.setTitle(this.getTitle(value[value.length - 1].data['title']));
                this.header = value[value.length - 1].data['title'];
                this.description = value[value.length - 1].data['description'];
            }
            this.changeDetectorRef.markForCheck();
        }));
        this.subscriptions.push(this.router.events.subscribe((routeEvent) => {
            if (routeEvent instanceof NavigationStart) {
                this.navigationEnd = false;
            }
            if (routeEvent instanceof NavigationEnd) {
                this.navigationEnd = true;
                this.setContentMinHeight();
            }
        }));
        this.subscriptions.push(this.layoutStore.sidebarLeftElementHeight.subscribe((value) => {
            this.sidebarLeftHeight = value;
            this.setContentMinHeight();
        }));
        this.subscriptions.push(this.layoutStore.layout.subscribe((value) => {
            this.layout = value;
            this.setContentMinHeight();
        }));
        this.subscriptions.push(this.layoutStore.windowInnerHeight.subscribe((value) => {
            this.windowInnerHeight = value;
            this.setContentMinHeight();
        }));
        this.heightStyle = this.windowInnerHeight;
    }
    /**
     * \@method ngOnDestroy
     * @return {?}
     */
    ngOnDestroy() {
        this.subscriptions = removeSubscriptions(this.subscriptions);
    }
    /**
     * [scrollHeight description]
     * \@method scrollHeight
     * @return {?} [description]
     */
    get scrollHeight() {
        return this.contentInnerElement.nativeElement.scrollHeight;
    }
    /**
     * [getTitle description]
     * \@method getTitle
     * @private
     * @param {?} title [description]
     * @return {?} [description]
     */
    getTitle(title) {
        return title ? `${title} - ${this.titleTag}` : this.titleTag;
    }
    /**
     * [setMinHeight description]
     * \@method setMinHeight
     * @private
     * @return {?}
     */
    setContentMinHeight() {
        if (this.navigationEnd) {
            /** @type {?} */
            let heightStyle;
            /** @type {?} */
            const headerFooterOffsetHeight = this.headerService.offsetHeight + this.footerService.offsetHeight;
            if (this.layout === 'fixed') {
                heightStyle = this.windowInnerHeight - this.footerService.offsetHeight;
            }
            else {
                /** @type {?} */
                const sidebarRight = this.sidebarRightService.scrollHeight ?
                    this.sidebarRightService.scrollHeight - this.headerService.offsetHeight : 0;
                heightStyle = Math.max(this.windowInnerHeight - headerFooterOffsetHeight, this.sidebarLeftHeight - this.headerService.offsetHeight, sidebarRight);
            }
            if (heightStyle && heightStyle !== this.heightStyle) {
                if (this.scrollHeight > heightStyle) {
                    heightStyle = null;
                }
                this.heightStyle = heightStyle;
                this.changeDetectorRef.detectChanges();
            }
        }
    }
}
ContentComponent.decorators = [
    { type: Component, args: [{
                selector: 'mk-layout-content',
                template: `<div class="content-wrapper" [style.min-height.px]="heightStyle">
  <div #contentInnerElement class="content-inner">
    <ng-content select="[mk-layout-content-before-header]"></ng-content>
    <section *ngIf="header || description" class="content-header">
      <h1>
        {{header}}
        <small *ngIf="description">{{description}}</small>
      </h1>
      <mk-breadcrumbs></mk-breadcrumbs>
    </section>
    <ng-content select="[mk-layout-content-after-header]"></ng-content>
    <section class="content">
      <ng-content></ng-content>
    </section>
  </div>
</div>
`,
                styles: [`:host{display:block}.content-wrapper{position:relative}`],
                changeDetection: ChangeDetectionStrategy.OnPush
            },] },
];
ContentComponent.ctorParameters = () => [
    { type: LayoutStore },
    { type: RoutingService },
    { type: Title },
    { type: ElementRef },
    { type: ChangeDetectorRef },
    { type: SidebarRightService },
    { type: HeaderService },
    { type: FooterService },
    { type: Router }
];
ContentComponent.propDecorators = {
    contentInnerElement: [{ type: ViewChild, args: ['contentInnerElement',] }]
};
if (false) {
    /** @type {?} */
    ContentComponent.prototype.description;
    /** @type {?} */
    ContentComponent.prototype.header;
    /** @type {?} */
    ContentComponent.prototype.heightStyle;
    /** @type {?} */
    ContentComponent.prototype.sidebarLeftHeight;
    /** @type {?} */
    ContentComponent.prototype.windowInnerHeight;
    /**
     * @type {?}
     * @private
     */
    ContentComponent.prototype.layout;
    /**
     * @type {?}
     * @private
     */
    ContentComponent.prototype.titleTag;
    /**
     * @type {?}
     * @private
     */
    ContentComponent.prototype.navigationEnd;
    /**
     * @type {?}
     * @private
     */
    ContentComponent.prototype.subscriptions;
    /**
     * @type {?}
     * @private
     */
    ContentComponent.prototype.contentInnerElement;
    /**
     * @type {?}
     * @private
     */
    ContentComponent.prototype.layoutStore;
    /**
     * @type {?}
     * @private
     */
    ContentComponent.prototype.routingService;
    /**
     * @type {?}
     * @private
     */
    ContentComponent.prototype.titleService;
    /**
     * @type {?}
     * @private
     */
    ContentComponent.prototype.elementRef;
    /**
     * @type {?}
     * @private
     */
    ContentComponent.prototype.changeDetectorRef;
    /**
     * @type {?}
     * @private
     */
    ContentComponent.prototype.sidebarRightService;
    /**
     * @type {?}
     * @private
     */
    ContentComponent.prototype.headerService;
    /**
     * @type {?}
     * @private
     */
    ContentComponent.prototype.footerService;
    /**
     * @type {?}
     * @private
     */
    ContentComponent.prototype.router;
}
